### PR TITLE
Allow to enabled ccm alpha features

### PR DESCRIPTION
**Pull request by dkistner**
**How to categorize this PR?**

/area open-source
/kind enhancement
/platform gcp

**What this PR does / why we need it**:
Allow to enable cloud controller manager alpha features. 
Alpha features can only be enabled when the `AllAlpha` feature flag is activated.

**Special notes for your reviewer**:
End to end test is missing therefore I keep it in draft state.

**Release note**:

```feature user
Allow to enable cloud-controller-manager alpha features via `cloudControllerManager.alphaFeatures[]` in the `ControlPlaneConfig. Alpha features can only be enabled when the `AllAlpha` feature flag is active.
```

/invite @prashanth26 @DockToFuture

### CODE REVIEW

**Review by vpnachev — COMMENTED** [review]
Could you implement some more enhanced validation like https://github.com/gardener/gardener/pull/4149? This way it will not be possible for the shoot owners to specify invalid feature gates that may prevent the CCM to start-up.

And in general, so far we never enabled alpha features because they can be removed at any time which makes the effort to enable it obsolete, but increases the operational burden.

**Comment by dkistner** [comment]
Generally I agree with you to not enable alpha features. I'm currently also not sure if we still need this PR as we had one stakeholder which require the `ILBCustomSubnet` because they need internal loadbalancers with ips from a specific subnet.
As the ccm in `v1.21` supports this without this alpha feature enabled and they start to adopt k8s `v1.21` we can probably also close this. Just removed the draft for completeness.
@stiller-leser Can you confirm?

File: pkg/apis/gcp/v1alpha1/types_controlplane.go [hunk 1]
```diff
@@ -38,4 +38,7 @@ type CloudControllerManagerConfig struct {
 	// FeatureGates contains information about enabled feature gates.
 	// +optional
 	FeatureGates map[string]bool `json:"featureGates,omitempty"`
+	// AlphaFeatureGates contains information about enabled alpha feature gates.
+	// Alpha Features require the "AllAlpha" feature gate enabled on the Cloud Controller Manager.
```
Comment (rfranzke) [inline pkg/apis/gcp/v1alpha1/types_controlplane.go:42]:
```suggestion
	// Alpha Features require the "AllAlpha" feature gate enabled on the Cloud Controller Manager.
	// +optional
```

File: pkg/controller/controlplane/valuesprovider.go [hunk 2]
```diff
@@ -417,9 +417,24 @@ func getConfigChartValues(
 		"subNetworkName": subNetworkName,
 		"zone":           cpConfig.Zone,
 		"nodeTags":       cp.Namespace,
+		"alphaFeatures":  getAlphaFeatuesValues(cpConfig.CloudControllerManager),
 	}, nil
 }
 
+func getAlphaFeatuesValues(ccmConfig *apisgcp.CloudControllerManagerConfig) []string {
+	var alphaFeatues = []string{}
```
Comment (rfranzke) [inline pkg/controller/controlplane/valuesprovider.go:425]:
```suggestion
	var alphaFeatues []string
```

File: pkg/controller/controlplane/valuesprovider.go [hunk 3]
```diff
@@ -417,9 +417,24 @@ func getConfigChartValues(
...
+
+	if ccmConfig == nil {
+		return alphaFeatues
+	}
```
Comment (rfranzke) [inline pkg/controller/controlplane/valuesprovider.go:429]:
```suggestion
	if ccmConfig == nil {
		return nil
	}
```

and move to the top?

**Review by rfranzke — COMMENTED** [review]
Minor feedback in case the PR is still relevant

**Comment by stiller-leser** [comment]
Hi all,

We do not need it anymore, but since it is almost done, wouldn't it be a useful addition?

Best regards

________________________________

File: charts/internal/cloud-provider-config/templates/cloud-provider-config.yaml [hunk 4]
```diff
@@ -15,3 +15,6 @@ data:
     local-zone="{{ .Values.zone }}"
     token-url=nil
     node-tags="{{ .Values.nodeTags }}"
+    {{- if .Values.alphaFeatures }}
+    alpha-features="{{range $index, $feature := .Values.alphaFeatures}}{{if $index}},{{end}}{{$feature}}{{end}}"
```
Comment (ialidzhikov) [inline charts/internal/cloud-provider-config/templates/cloud-provider-config.yaml:19]: Ha, I didn't know that gcp CCM supports such config option. It is confusing and not what I would expect honestly. CCMs have a `--feature-gates` flag and I would expect even out-of-tree provider-specific features to be configured via this flag.

File: docs/usage-as-end-user.md [hunk 5]
```diff
@@ -123,6 +123,9 @@ The `cloudControllerManager.featureGates` contains a map of explicitly enabled o
 For production usage it's not recommend to use this field at all as you can enable alpha features or disable beta/stable features, potentially impacting the cluster stability.
 If you don't want to configure anything for the `cloudControllerManager` simply omit the key in the YAML specification.
 
+The `cloudControllerManager.alphaFeatureGates` contains a map of enabled or disabled feature gates.
+The alpha features can only be enabled in when the `AllAlpha` feature gate is enabled.
```
Comment (ialidzhikov) [inline docs/usage-as-end-user.md:127]: Isn't `AllAlpha` too dangerous? I guess currently there are alpha features that can cause issues when they are enabled. One good example is the `MemoryQoS` feature gate that requires (I expect) cgroup v2 that is currently not supported (disabled) by gardenlinux - https://github.com/gardenlinux/gardenlinux/pull/205.
I guess no one of us would like to dig for issue in a cluster where AllAlpha feature gate is enabled.

**Comment by dkistner** [comment]
As said, this was intended to unblock @stiller-leser use case, but as they now have a solution by with moving their clusters to k8s `v1.21` that does not require the alpha feature anymore I would also tend to not allowing enabling alpha features.
So I would close this PR. In case we need it we can reactive again.
/close

**Pull request closed without merging**